app/service/form_data_test.go

```diff
@@ -0,0 +1,210 @@
+		},
+		{
+			"decoder error for a field",
+			&struct {
+				Time time.Time `json:"time"`
+			}{},
+			`{"time":"123"}`,
+			"invalid input data",
+			FieldErrors{"time": {"decoding error: parsing time \"123\" as \"2006-01-02T15:04:05Z07:00\": cannot parse \"123\" as \"2006\""}},
+		},
+		{
+			"multiple errors",
+			&struct {
+				ID int64 `json:"id" valid:"required"`
+			}{},
+			`{"my_id":1, "id":0}`,
+			"invalid input data",
+			FieldErrors{
+				"id":    {"non zero value required"},
+				"my_id": {"unexpected field"},
+			},
+		},
+		{
+			"nested structure",
+			&struct {
+				Struct struct {
+					Name        string `json:"name" valid:"required"`
+					OtherStruct struct {
+						Name string `json:"name" valid:"required"`
+					} `json:"other_struct" valid:"required"`
+				} `json:"struct" valid:"required"`
+			}{},
+			`{"id":0}`,
+			"invalid input data",
+			FieldErrors{
+				"id":                      {"unexpected field"},
+				"struct":                  {"non zero value required"},
+				"Struct.name":             {"non zero value required"},
+				"Struct.other_struct":     {"non zero value required"},
+				"Struct.OtherStruct.name": {"non zero value required"},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			f := NewFormData(tt.definitionStructure)
+			req, _ := http.NewRequest("POST", "/", strings.NewReader(tt.json))
+			err := f.ParseJSONRequestData(req)
+			if tt.wantErr != "" {
+				assert.Equal(t, tt.wantErr, err.Error())
+			} else {
+				assert.Nil(t, err)
+			}
+			if tt.wantFieldErrors != nil {
+				assert.IsType(t, FieldErrors{}, err)
+				assert.Equal(t, tt.wantFieldErrors, err)
+			}
+		})
+	}
+}
+
+func TestFormData_ConstructMapForDB(t *testing.T) {
+	tests := []struct {
+		name                string
+		definitionStructure interface{}
+		json                string
+		want                map[string]interface{}
+	}{
+		{
+			"simple",
+			&struct {
+				ID   int64  `json:"id"`
+				Name string `json:"name"`
+			}{},
+			`{"id":123, "name":"John"}`,
+			map[string]interface{}{"ID": int64(123), "Name": "John"},
+		},
+		{
+			"sql and gorm tags",
+			&struct {
+				ID    int64  `json:"id" sql:"column:id"`
+				Name  string `json:"name" gorm:"column:sName"`
+				Skip  string `json:"skip1" gorm:"-"`
+				Skip2 string `json:"skip2" sql:"-"`
+			}{},
+			`{"id":123, "name":"John", "skip1":"skip", "skip2":"skip"}`,
+			map[string]interface{}{"id": int64(123), "sName": "John"},
+		},
+		{
+			"skips missing fields",
+			&struct {
+				Name        string `json:"name" gorm:"column:sName"`
+				Description string `json:"description" gorm:"column:sDescription"`
+			}{},
+			`{}`,
+			map[string]interface{}{},
+		},
+		{
+			"skips unexported fields",
+			&struct {
+				name string
+			}{},
+			`{"name":"test"}`,
+			map[string]interface{}{},
+		},
+		{
+			"skips fields ignored by json",
+			&struct {
+				Name string `json:"-" sql:"column:sName"`
```

Comment:
If I provide this field in JSON, I get `service.FieldErrors{"Name":[]string{"unexpected field"}}`

Reply:
`ConstructMapForDB()` works with the structure. It knows nothing about the input JSON. Since we have an exported field `Name`, `ConstructMapForDB()` will set its value in the map. So this test makes sense.